---
title: "Set up combined data of Mutect2 and Strelka2"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

Candace Savonen - CCDL for ALSF

This notebook sets up the [MAF data files as a combined file for ready comparison in 02-analyze-concordance.Rmd](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/mutect2-vs-strelka2/02-analyze-concordance.Rmd)
and also does some first line analyses from ready-made `maftools` functions.

It is the first notebook in this series which addresses [issue \# 30 in OpenPBTA](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/30).

### Summary of the Set Up:  

For both the Strelka2 and MuTect2 datasets, the data is imported by 
`maftools::read.maf` and the corresponding clinical data from 
`pbta-histologies.tsv` is added to the object. 
This is only done once as written (as the `read.maf` is very memory intensive)
and each MuTect2 and Strelka2 are saved to an RDS file for faster and ready 
reloading.

For both Strelka2 and MuTect2 data, the following variables are calculated
and added into the combined dataset for analysis in [02-analyze-concordance.Rmd](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/mutect2-vs-strelka2/02-analyze-concordance.Rmd).

#### Variables created:  
- `vaf` : Variant Allele Frequency = `(t_alt_count) / (t_ref_count + t_alt_count)`.  
- `mutation_id`: Used to determine whether two variants calls are identical 
in both datasets. 
It is a concatenation of: `Hugo_Symbol`, `change`, `Start_Position`, and 
`Tumor_Sample_Barcode` (the sample ID).   
- `base_change`: variable that indicates the exact change in bases (e.g. 'T>C').  
- `change`: indicates the `base_change` information but groups together 
deletions, insertions, and long (more than a SNV) as their own groups.  
- `coding`: Summarize the `BIOTYPE` variable for whether or not it is a coding gene.  

#### The output files from this notebook:
- `scratch/strelka2.RDS`  
- `scratch/mutect2.RDS`  
- `scratch/metadata_filtered_maf_samples.tsv`  
- `analyses/mutect2-vs-strelka2/plots/sample_cor_mutect2_vs_strelka2.pdf`
- `analyses/mutect2-vs-strelka2/plots/sample_cor_mutect2_vs_strelka2.pdf`
- `analyses/mutect2-vs-strelka2/results/combined_results.tsv`  

## Usage

To run this from the command line, use:
```
Rscript -e "rmarkdown::render('analyses/mutect2-vs-strelka2/01-set-up.Rmd', 
                              clean = TRUE)"
```
 _This assumes you are in the top directory of the repository._

## Set Up

```{r}
# We need maftools - this will be added to the running Docker issue whenever it is up
if (!("maftools" %in% installed.packages())) {
  if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
  }
  BiocManager::install("maftools")
}
```

Get `magrittr` pipe

```{r}
`%>%` <- dplyr::`%>%`
```

### Directories and files

Path to the symlinked data obtained via `bash download-data.sh`.

```{r}
data_dir <- file.path("..", "..", "data")
scratch_dir <- file.path("..", "..", "scratch")
```

Create output directories in this analysis folder.

```{r}
if (!dir.exists("results")) {
  dir.create("results")
}
if (!dir.exists("plots")) {
  dir.create("plots")
}
```

## Read in the metadata information

Running `maftools::read.maf` takes a lot of computing power and time, so to 
avoid having to run this for both datasets everytime we want to re-run this 
notebook or the analyses in the other notebook, I've set this up to save the 
`MAF` objects as `RDS` files.

First let's establish the file paths.

```{r}
# File paths for the needed files for this analysis
metadata_dir <- file.path(scratch_dir, "metadata_filtered_maf_samples.tsv")
strelka2_dir <- file.path(scratch_dir, "strelka2.RDS")
mutect2_dir <- file.path(scratch_dir, "mutect2.RDS")
```

## Read in the Strelka2 and Mutect2 data

Will read in the data as `maftools` objects from an RDS file, unless `maftools` 
has not been run on them yet.
We will use `metadata` as the `clinicalData` for the `maftools` object. 
 
Note: If you trying to run the initialset up step in a Docker container, it will likely be 
out of memory killed, unless you have ~50GB you can allot to Docker. 

```{r}
# Get a vector of whether these exist
files_needed <- file.exists(metadata_dir, strelka2_dir, mutect2_dir)

if (all(files_needed)) {
  # Read the ready-to-go files if these files exist
  metadata <- metadata <- readr::read_tsv(metadata_dir)
  strelka2 <- readRDS(strelka2_dir)
  mutect2 <- readRDS(mutect2_dir)
} else { # If any of the needed files don't exist, rerun this process:
  # Only import the sample names
  strelka2_samples <- data.table::fread(file.path(
    data_dir,
    "pbta-snv-strelka2.vep.maf.gz"
  ),
  select = "Tumor_Sample_Barcode",
  skip = 1,
  data.table = FALSE
  )

  mutect2_samples <- data.table::fread(file.path(
    data_dir,
    "pbta-snv-mutect2.vep.maf.gz"
  ),
  select = "Tumor_Sample_Barcode",
  skip = 1,
  data.table = FALSE
  )

  # Isolate metadata to only the samples that are in the datasets
  metadata <- readr::read_tsv(data_dir, "pbta-histologies.tsv") %>%
    dplyr::filter(Kids_First_Biospecimen_ID %in% c(strelka2_samples, mutect2_samples)) %>%
    dplyr::distinct(Kids_First_Biospecimen_ID, .keep_all = TRUE) %>%
    dplyr::arrange() %>%
    readr::write_tsv(file.path(scratch_dir, "metadata_filtered_maf_samples.tsv"))

  # Read in original strelka file with maftools
  strelka <- maftools::read.maf(file.path(data_dir, "pbta-snv-strelka2.vep.maf.gz"),
    clinicalData = metadata
  )

  # Save to RDS so we don't have to run this again
  saveRDS(strelka, strelka2_dir)

  # Same for MuTect2
  mutect2 <- maftools::read.maf(file.path(data_dir, "pbta-snv-mutect2.vep.maf.gz"),
    clinicalData = metadata
  )
  saveRDS(mutect2, mutect2_dir)
}
```

## Get summaries and write them to TSVs 

Get gene summaries and write to TSV files. 

```{r}
strelka2_gene_sum <- maftools::getGeneSummary(strelka2) %>%
  readr::write_tsv(file.path(
    "results",
    "strelka2_gene_summary.tsv"
  ))

mutect2_gene_sum <- maftools::getGeneSummary(mutect2) %>%
  readr::write_tsv(file.path(
    "results",
    "mutect2_gene_summary.tsv"
  ))
```

Get sample summaries and write to TSV files. 

```{r}
strelka2_sample_sum <- maftools::getSampleSummary(strelka2) %>%
  readr::write_tsv(file.path(
    "results",
    "strelka2_sample_summary.tsv"
  ))

mutect2_sample_sum <- maftools::getSampleSummary(mutect2) %>%
  readr::write_tsv(file.path(
    "results",
    "mutect2_sample_summary.tsv"
  ))
```

## Number of mutations per gene correlation 

```{r}
combined_gene <- mutect2_gene_sum %>%
  dplyr::full_join(strelka2_gene_sum, by = "Hugo_Symbol") %>%
  reshape2::melt(id = "Hugo_Symbol") %>%
  dplyr::mutate(dataset = as.character(grepl(".x$", variable))) %>%
  dplyr::mutate(dataset = dplyr::recode(dataset,
    `TRUE` = "mutect2",
    `FALSE` = "strelka2"
  )) %>%
  dplyr::mutate(variable = gsub(".x$|.y$", "", variable)) %>%
  tidyr::spread("dataset", "value")
```

Let's get a correlation test on the genes overall.

```{r}
cor.test(combined_gene$mutect2, combined_gene$strelka2, method = "spearman")
cor.test(combined_gene$mutect2, combined_gene$strelka2, method = "pearson")
```

## Number of mutations per sample correlation. 

```{r}
combined_sample <- mutect2_sample_sum %>%
  dplyr::full_join(strelka2_sample_sum, by = "Tumor_Sample_Barcode") %>%
  reshape2::melt(id = "Tumor_Sample_Barcode") %>%
  dplyr::mutate(dataset = as.character(grepl(".x$", variable))) %>%
  dplyr::mutate(dataset = dplyr::recode(dataset,
    `TRUE` = "mutect2",
    `FALSE` = "strelka2"
  )) %>%
  dplyr::mutate(variable = gsub(".x$|.y$", "", variable)) %>%
  tidyr::spread("dataset", "value")
```

Let's get a correlation test on the genes overall.

```{r}
cor.test(combined_sample$mutect2, combined_sample$strelka2, method = "spearman")
cor.test(combined_sample$mutect2, combined_sample$strelka2, method = "pearson")
```

## Plot Transition/Transversions

```{r}
maftools::plotTiTv(maftools::titv(strelka2))
```

```{r}
maftools::plotTiTv(maftools::titv(mutect2))
```

## Set up new variables

Here we will make these new variables for both Mutect2 and Strelka2 dataset:
- Calculate VAF for each. 
The VAFs for each variation in each dataset are calculated by 
`t_alt_count) / (t_ref_count + t_alt_count) `, as following the [code used in 
`maftools`](https://github.com/PoisonAlien/maftools/blob/master/R/plot_vaf.R)  
- Create a `base_change` variable that indicates the exact change in bases.  
- Create a `change` variable for each which indicates the `base_change`
information but groups together deletions, insertions, and long (more than a 
SNV) as their own groups.  
- Make a mutation id by concatenating `Hugo_Symbol`, `change`, `Start_Position`,
and `Tumor_Sample_Barcode` (the sample ID).   
We will use this variable to determine whether two variants calls are identical 
in both datasets. 
- Summarize the `BIOTYPE` variable for whether or not it is a coding gene.   

Let's set up Strelka2's variables first. 

```{r}
strelka2_vaf <- strelka2@data %>%
  dplyr::mutate(
    vaf = as.numeric(t_alt_count) / (as.numeric(t_ref_count) +
      as.numeric(t_alt_count)),
    base_change = paste0(Reference_Allele, ">", Allele),
    coding = dplyr::case_when(
      BIOTYPE != "protein_coding" ~ "non-coding",
      TRUE ~ "protein_coding"
    )
  ) %>%
  dplyr::mutate(change = dplyr::case_when(
    grepl("^-", base_change) ~ "insertion",
    grepl("-$", base_change) ~ "deletion",
    nchar(base_change) > 3 ~ "long_change",
    TRUE ~ base_change
  )) %>%
  dplyr::mutate(
    mutation_id = paste0(
      Hugo_Symbol, "_",
      change, "_",
      Start_Position, "_",
      Tumor_Sample_Barcode
    )
  ) %>%
  dplyr::select(-which(apply(is.na(.), 2, all)))

# Take a look at this df
strelka2_vaf
```

Now we will do the same for MuTect2.

```{r}
mutect2_vaf <- mutect2@data %>%
  dplyr::mutate(
    vaf = as.numeric(t_alt_count) / (as.numeric(t_ref_count) +
      as.numeric(t_alt_count)),
    base_change = paste0(Reference_Allele, ">", Allele),
    coding = dplyr::case_when(
      BIOTYPE != "protein_coding" ~ "non-coding",
      TRUE ~ "protein_coding"
    )
  ) %>%
  dplyr::mutate(change = dplyr::case_when(
    grepl("^-", base_change) ~ "insertion",
    grepl("-$", base_change) ~ "deletion",
    nchar(base_change) > 3 ~ "long_change",
    TRUE ~ base_change
  )) %>%
  dplyr::mutate(
    mutation_id = paste0(
      Hugo_Symbol, "_",
      change, "_",
      Start_Position, "_",
      Tumor_Sample_Barcode
    )
  ) %>%
  dplyr::select(-which(apply(is.na(.), 2, all)))

# Take a look at this df
mutect2_vaf
```

## Combine MuTect2 and Strelka2 data.frames into one data.frame

Join these datasets based on the `mutation_id` created above. 
Save to a TSV file to be used in the subsequent notebook.

```{r}
# Merge these data.frames together
vaf_df <- strelka2_vaf %>%
  dplyr::full_join(mutect2_vaf,
    by = "mutation_id",
    suffix = c(".strelka2", ".mutect2")
  ) %>%
  # Make a variable that denotes which dataset it is in.
  dplyr::mutate(dataset = dplyr::case_when(
    is.na(Allele.mutect2) ~ "strelka2_only",
    is.na(Allele.strelka2) ~ "mutect2_only",
    TRUE ~ "both"
  )) %>%
  readr::write_tsv(file.path("results", "combined_results.tsv"))
```

Make a zipped up version that can be stored on GitHub.

```{r}
zip(file.path("results", "combined_results.tsv.zip"),
    file.path("results", "combined_results.tsv"))
```

Session Info: 

```{r}
sessionInfo()
```
