---
title: "Evaluate concordance between Mutect2 and Strelka2"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

Candace Savonen - CCDL for ALSF

This notebook is a first pass in evaluating the concordance between MuTect2 and 
Strelka2 results.

It addresses [issue \# 30 in OpenPBTA](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/30).

## Set Up

\* Some of these set up steps will be removed once we have a Dockerfile that installs
maftools automatically. 

```{r}
# We need maftools - this will be added to the running Docker issue whenever it is up
if (!('maftools' %in% installed.packages())) {
  devtools::install_github("PoisonAlien/maftools")
}

if (!('hexbin' %in% installed.packages())) {
  install.packages("hexbin")
}

if (!('colorblindr' %in% installed.packages())) {
  devtools::install_github("clauswilke/colorblindr")
}
# Get magrittr pipe
`%>%` <- dplyr::`%>%`
```

Create a output directories.

```{r}
if (!dir.exists("results")) {
  dir.create("results")
}
if (!dir.exists("plots")) {
  dir.create("plots")
}
```

## Read in the metadata information
This will be changed once the [data download script](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/2) is finalized. 

Prep the metadata to be used as the `clinicalData` for maftools it it hasn't been 
prepped yet. 
This whole chunk may need to be removed be taken out after issue 2 regarding the [data download script](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/2) is addressed. 

These steps are only to make the metadata and MuTect2 and Strelka2 datasets 
parallel in what samples they contain.  
There are some samples in the metadata that are not in the MuTect2 and Strelka2
data. 

```{r}
if (file.exists(file.path("..", "..", "scratch", "metadata_filtered.tsv"))) {
  metadata <- readr::read_tsv(file.path("..", "..", "scratch", "metadata_filtered.tsv"))
}
```

## Read in the Strelka2 and Mutect2 data

Will read in as an `maftools` object from an RDS file, unless `maftools` has not been 
run on them yet.
Running `maftools` takes a lot of computing power and time, so be prepared.
If you trying to run this step in a Docker container, it will likely be out of 
memory killed, unless you have ~50GB you can allot to Docker. 

```{r}
# Load in the RDS file if it exists, but if it doesn't exist, load in from 
# original data file with maftools
if (!file.exists(file.path("..", "..", "scratch", "strelka2.RDS"))) {
  strelka <- maftools::read.maf(file.path("..", "..", "data", "strelka2.maf.gz"),
                                clinicalData = metadata)
  saveRDS(strelka, file.path("..", "..", "scratch", "strelka2.RDS"))
} else {
  strelka2 <- readRDS(file.path("..", "..", "scratch", "strelka2.RDS"))
}

# Same for MuTect2
if (!file.exists(file.path("..", "..", "scratch", "mutect2.RDS"))) {
  mutect2 <- maftools::read.maf(file.path("..", "..", "data", "mutect2.maf.gz"), 
                                clinicalData = metadata)
  saveRDS(mutect2, file.path("..", "..", "scratch", "mutect2.RDS"))
} else {
  mutect2 <- readRDS(file.path("..", "..", "scratch", "mutect2.RDS"))
}
```

This is how I set up `metadata_filtered.tsv`. 
This should not need to be ran again and won't run again unless 
`metadata_filtered.tsv` is misplaced from the scratch file. 

```{r}
if (!file.exists(file.path("..", "..", "scratch", "metadata_filtered.tsv"))) {
  # Isolate metadata to only the samples that are in the datasets
  metadata <- metadata %>%
    dplyr::filter(sample_id %in% mutect2@clinical.data$Tumor_Sample_Barcode) %>%
    dplyr::distinct(sample_id, .keep_all = TRUE) %>%
    readr::write_tsv(file.path("..", "..", "scratch", "metadata_filtered.tsv"))
}
```

Check that samples are same order in the datasets as they are in the metadata

```{r}
all.equal(as.factor(metadata$sample_id),
          strelka2@clinical.data$Tumor_Sample_Barcode)

all.equal(as.factor(metadata$sample_id), 
          mutect2@clinical.data$Tumor_Sample_Barcode)
```

## Get summaries and write them to TSVs 

Get gene summaries and write to TSV files. 

```{r}
strelka2_gene_sum <- maftools::getGeneSummary(strelka2) %>% 
  readr::write_tsv(file.path("results", 
                             "strelka2_gene_summary.tsv"))

mutect2_gene_sum <- maftools::getGeneSummary(mutect2) %>% 
  readr::write_tsv(file.path("results", 
                             "mutect2_gene_summary.tsv"))
```

Get sample summaries and write to TSV files. 

```{r}
strelka2_sample_sum <- maftools::getSampleSummary(strelka2) %>% 
  readr::write_tsv(file.path("results", 
                             "strelka2_sample_summary.tsv"))

mutect2_sample_sum <- maftools::getSampleSummary(mutect2) %>% 
  readr::write_tsv(file.path("results", 
                             "mutect2_sample_summary.tsv"))
```

## Number of mutations per gene correlation 

```{r}
combined_gene <- mutect2_gene_sum %>% 
  dplyr::full_join(strelka2_gene_sum, by = 'Hugo_Symbol') %>%
  reshape2::melt(id = 'Hugo_Symbol') %>% 
  dplyr::mutate(dataset = as.character(grepl(".x$", variable))) %>%
  dplyr::mutate(dataset = dplyr::recode(dataset, 
                                        `TRUE` = "mutect2", 
                                        `FALSE` = "strelka2")) %>%
  dplyr::mutate(variable = gsub(".x$|.y$", "", variable)) %>% 
  tidyr::spread('dataset', 'value')
```

Make number of mutations per gene scatterplots.

```{r}
gene_cor <- ggplot2::ggplot(combined_gene, ggplot2::aes(x = mutect2, y = strelka2)) +
  ggplot2::geom_hex(bins = 10) + 
  ggplot2::facet_wrap(~variable, scales = "free") + 
  ggplot2::geom_smooth(method = lm) + 
  ggplot2::theme_classic() + 
  ggplot2::xlab("Mutect2: Number of mutations per gene") +
  ggplot2::ylab("Strelka2: Number of mutations per gene") 

# Print out the plot in this notebook
gene_cor
```

Save the plot to a PDF.

```{r}
ggplot2::ggsave(file.path("plots", "gene_cor_mutect2_vs_strelka2.pdf"))
```

Let's get a correlation test on the genes overall.

```{r}
cor.test(combined_gene$mutect2, combined_gene$strelka2, method = "spearman")
cor.test(combined_gene$mutect2, combined_gene$strelka2, method = "pearson")
```

## Number of mutations per sample correlation. 

```{r}
combined_sample <- mutect2_sample_sum %>% 
  dplyr::full_join(strelka2_sample_sum, by = 'Tumor_Sample_Barcode') %>%
  reshape2::melt(id = 'Tumor_Sample_Barcode') %>% 
  dplyr::mutate(dataset = as.character(grepl(".x$", variable))) %>%
  dplyr::mutate(dataset = dplyr::recode(dataset, 
                                        `TRUE` = "mutect2", 
                                        `FALSE` = "strelka2")) %>%
  dplyr::mutate(variable = gsub(".x$|.y$", "", variable)) %>% 
  tidyr::spread('dataset', 'value')
```

Make number of mutations per sample scatterplots.

```{r}
sample_cor <- ggplot2::ggplot(combined_sample, ggplot2::aes(x = mutect2, y = strelka2)) +
  ggplot2::geom_hex(bins = 10) + 
  ggplot2::facet_wrap(~variable, scales = "free") + 
  ggplot2::geom_smooth(method = lm) + 
  ggplot2::theme_classic() +
  ggplot2::xlab("Mutect2: Number of mutations per sample") +
  ggplot2::ylab("Strelka2: Number of mutations per sample") 

# Print out the plot in this notebook
sample_cor
```

Save the plot to a PDF.

```{r}
ggplot2::ggsave(file.path("plots", "sample_cor_mutect2_vs_strelka2.pdf"))
```

Let's get a correlation test on the genes overall.
**Question 2)** Is this a reasonable amount of concordance for these methods 
across samples? 

```{r}
cor.test(combined_sample$mutect2, combined_sample$strelka2, method = "spearman")
cor.test(combined_sample$mutect2, combined_sample$strelka2, method = "pearson")
```

## Calculate VAF

Set up data.frame with mutation and VAF for Strelka2.

```{r}
strelka2_vaf <- strelka2@data %>%
  dplyr::filter(!grepl("-", Allele)) %>%
  dplyr::mutate(vaf = as.numeric(t_alt_count)/(as.numeric(t_ref_count) + 
                                                 as.numeric(t_alt_count)),
                mutation = paste0(Hugo_Symbol, "_", 
                                  Allele, "_", 
                                  Tumor_Sample_Barcode, "_", 
                                  Start_Position), 
                change = paste0(Reference_Allele, ">", Allele),
                coding = dplyr::case_when( 
                  BIOTYPE != "protein_coding" ~ "non-coding",
                  TRUE ~ "protein_coding")) %>%
    dplyr::select(-which(apply(is.na(.), 2, all)))

# Take a look at this df
strelka2_vaf
```

Set up data.frame with mutation and VAF for MuTect2.

```{r}
mutect2_vaf <- mutect2@data %>%
  dplyr::filter(!grepl("-", Allele)) %>%
  dplyr::mutate(vaf = as.numeric(t_alt_count)/(as.numeric(t_ref_count) + 
                                                 as.numeric(t_alt_count)),
                mutation = paste0(Hugo_Symbol, "_", 
                                  Allele, "_", 
                                  Tumor_Sample_Barcode, "_", 
                                  Start_Position),
                change = paste0(Reference_Allele, ">", Allele), 
                coding = dplyr::case_when( 
                  BIOTYPE != "protein_coding" ~ "non-coding",
                  TRUE ~ "protein_coding")) %>%
  dplyr::select(-which(apply(is.na(.), 2, all)))

# Take a look at this df
mutect2_vaf
```

Combine MuTect2 and Strelka2 VAF data.frames so we can compare.

```{r}
# Merge these data.frames together
vaf_df <- strelka2_vaf %>%
  dplyr::full_join(mutect2_vaf, by = 'mutation', 
                    suffix = c(".strelka2", ".mutect2")) %>%
  # Make a variable that denotes which dataset it is in.
  dplyr::mutate(dataset = dplyr::case_when( 
    is.na(vaf.mutect2) ~ "strelka2_only",
    is.na(vaf.strelka2) ~ "mutect2_only", 
    TRUE ~ "both")) 
```

Plot this as a scatterplot

```{r}
vaf_df %>%
ggplot2::ggplot(ggplot2::aes(x = vaf.strelka2, y = vaf.mutect2)) + 
  ggplot2::geom_hex() +
  ggplot2::theme_classic() + 
  ggplot2::xlab("VAF for each mutation for Strelka2") + 
  ggplot2::ylab("VAF for each mutation for MuTect2") 
```

```{r}
vaf_df %>%
  tidyr::gather(key = "data", value = "vaf" , vaf.strelka2, vaf.mutect2) %>%
  dplyr::mutate(data = gsub("^vaf.", "", data)) %>% 
  dplyr::mutate(data.group = paste(dataset, ":", data, "VAF")) %>%
  dplyr::filter(!is.na(vaf)) %>%
# Plot it
ggplot2::ggplot(ggplot2::aes(data.group, vaf)) + 
  ggplot2::geom_violin(fill = "light blue") +
  ggplot2::theme_classic( ) + 
  ggplot2::ylab("Density of VAF") + 
  ggplot2::xlab(" ")
```

## Venn Diagrams

Make the Venn diagram of MuTect2 and Strelka2 mutations. 

```{r}
count <- summary(as.factor(vaf_df$dataset))

# Take a look at this summary
count
```

```{r}
# Make the Venn diagram
grid::grid.newpage();
venn.plot <- VennDiagram::draw.pairwise.venn(
  area1 = count[3] + count[1],
  area2 = count[2] + count[1],
  cross.area = count[1],
  category = c("Strelka2", "MuTect2"),
  fill = c("blue", "yellow"),
  cex = 2,
  cat.cex = 1.5,
  cat.dist = c(-0.04, -0.031),
  ext.pos = 0,
  ext.dist = -0.01,
  ext.length = .8,
  ext.line.lwd = 2,
  ext.line.lty = "dashed");
grid::grid.draw(venn.plot) # Draw plot
```

Save the Venn Diagram plot to a PDF. 

```{r}
# Make filename to save plot as
venn.plot.file <- file.path("plots", 
                            "strelka2_mutect2_venn_diagram.pdf")
pdf(venn.plot.file);
grid::grid.draw(venn.plot);
dev.off()
```

## What types of variants are are the most discrepant between the algorithms?

Let's make a wrapper function that will do this for whatever variables we are 
interested in. 

```{r}
barplot_var <- function(df, variable_name = NULL, filter_cutoff = 1, 
                        as_percent = FALSE) {
  count_df <- vaf_df %>%
    dplyr::select(paste0(variable_name, c(".strelka2", ".mutect2")),
                  dataset) %>%
    tidyr::gather(key = "data", value = "variable" , 
                  paste0(variable_name, c(".strelka2", ".mutect2"))) %>%
    dplyr::mutate('data' = gsub(paste0(variable_name, "."), "", data)) 
  
  %>%
    dplyr::group_by(dataset, data, variable) %>% 
    dplyr::summarise(count = n())
  
  totals <- count_df %>%
    dplyr::group_by(dataset) %>%
    dplyr::summarise(total = sum(count, na.rm = TRUE)) 
    
    
  count_df %>%
    dplyr::mutate(percent = count_df$count/totals$total[match(count_df$dataset, totals$dataset)])
                                  
    dplyr::filter(count > filter_cutoff, !is.na(variable)) %>%
    dplyr::mutate(report = ifelse(as_percent, percent, count)) %>%
    ggplot2::ggplot(ggplot2::aes(x = reorder(variable, -report), 
                                 y = report, fill = dataset)) + 
    ggplot2::geom_bar(position = "dodge", stat="identity") + 
    ggplot2::theme_classic() +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, hjust = 1)) +
    colorblindr::scale_fill_OkabeIto() + 
    ggplot2::xlab("")
}
```

```{r}
barplot_var(vaf_df, "change", 100)
```

```{r}
barplot_var(vaf_df, "coding", 10)
```

```{r}
barplot_var(vaf_df, "BIOTYPE", 10)
```

```{r}
barplot_var(vaf_df, "Variant_Classification", 10)
```

```{r}
barplot_var(vaf_df, "VARIANT_CLASS", 10)
```

```{r}
barplot_var(vaf_df, "IMPACT", 10)
```

Session Info: 

```{r}
sessionInfo()
```